[KOGITO-3642] - SW Example with Events, REST Functions and RuleUnits (#448)
kiegroup/kogito-examples: serverless-workflow-functions-events-quarkus/README.md

# Kogito Serverless Workflow - Patient Onboarding

## Description

In this example we will see how to use the [Serverless Workflow specification](https://github.com/serverlessworkflow/specification)
implementation on Kogito to orchestrate a sequence of REST calls in an Event-Driven architecture.

This use case is about onboarding a new patient in a hospital, and based on their symptom, the 
application will assign to the right expert. 

The workflow starts with a CloudEvent informing a new patient entry:

```json

> {

Reviewer: @ricardozanini would be nice to also have the CURL command example here so people can simply copy and execute, like https://github.com/kiegroup/kogito-examples/tree/stable/kogito-travel-agency/extended/travels#post-travels

# Kogito Serverless Workflow - Patient Onboarding

## Description

In this example we will see how to use the [Serverless Workflow specification](https://github.com/serverlessworkflow/specification)
implementation on Kogito to orchestrate a sequence of REST calls in an Event-Driven architecture.

This use case is about onboarding a new patient in a hospital, and based on their symptom, the 
application will assign to the right expert. 

The workflow starts with a CloudEvent informing a new patient entry:

```json
{
   "id":"12345",
   "source":"/hospital/entry",
   "specversion":"1.0",
   "type":"new.patient.events",
   "data":{
      "name":"John",
      "dateOfBirth":"2017-08-15",
      "symptoms":[
         "seizures"
      ]
   }
}
```

> Or you can use CURL and set the CloudEvents parameters in the request headers:

Reviewer 2: ```suggestion
Or you can use cURL and set the CloudEvents parameters in the request headers:
```

```shell script
$ curl -X POST \
      -H "content-type: application/json"  \
      -H "ce-specversion: 1.0"  \
      -H "ce-source: /hospital/entry"  \
      -H "ce-type: new.patients.events"  \
      -H "ce-id: 12346"  \
      -d "{ \"name\": \"Mick\", \"dateOfBirth\": \"1983-08-15\", \"symptoms\":[\"seizures\"]}" \
  http://localhost:8080
```

The [workflow](src/main/resources/onbording.sw.json) will then call a sequence of REST services:

> 1. `StoreNewPatient` will store the patient in the internal database for later processing, generating a new ID for they

Reviewer 2: `generating a new ID for they` ???

> 2. `AssignDoctorToPatient` service will run a set of [rules](https://docs.jboss.org/kogito/release/latest/html_single/#con-drl-rule-units_drl-rules) 

Reviewer 2: ```suggestion
2. `AssignDoctorToPatient` will run a set of [rules](https://docs.jboss.org/kogito/release/latest/html_single/#con-drl-rule-units_drl-rules) 
```
for consistency in phrasing

> to determine which doctor this patient should be assigned based on their symptoms or general characteristics (children always go to Pediatrics, for example)

Reviewer 2: ```suggestion
to determine to which doctor this patient should be assigned, based on their symptoms or general characteristics (children always go to Pediatrics, for example)
```

> 3. `SchedulePatientAppointment` will find a schedule for this patient based on the agenda of the doctor assigned

Reviewer 2: ```suggestion
3. `SchedulePatientAppointment` will find a time slot for this patient based on the agenda of the assigned doctor
```

After executing the workflow, the appointments' data can be fetched using the `/onboarding/schedule/appointment` endpoint.

There's a nice web interface for you to try the example available in the root url (http://localhost:8080 for local environments):

![](imgs/web_ui.png)

Enter the data into the provided form to see the example in action! Date of Birth must be in `YYYY-MM-DD` format.

> Also, there's the Swagger UI that can also be used to explore the APIs in this example: http://localhost:8080/swagger-ui/

Reviewer 2: ```suggestion
Also, there is the Swagger UI that can also be used to explore the APIs in this example: http://localhost:8080/swagger-ui/
```

## Installing and Running

### Prerequisites
 
You will need:
  - Java 11+ installed
  - Environment variable JAVA_HOME set accordingly
  - Maven 3.6.2+ installed

When using native image compilation, you will also need: 
  - [GraalVm](https://www.graalvm.org/downloads/) 20.2.0+ installed
  - Environment variable GRAALVM_HOME set accordingly

>   - Note that GraalVM native image compilation typically requires other packages (glibc-devel, zlib-devel and gcc) to be installed too.  You also need 'native-image' installed in GraalVM (using 'gu install native-image'). Please refer to [GraalVM installation documentation](https://www.graalvm.org/docs/reference-manual/aot-compilation/#prerequisites) for more details.

Reviewer 2: thinking that now, with https://quarkus.io/guides/building-native-image#container-runtime, you don't need Gradle installed anymore if you have Docker/Podman ;)

Author: Well, I believe we need to rephrase all the other examples as well! :sweat:

### Compile and Run in Local Dev Mode

```text
mvn clean package quarkus:dev    
```

> Use `curl` to send the CloudEvent through HTTP to the application:

Reviewer 2: ```suggestion
Use `curl` command to send the CloudEvent through HTTP to the application:
```

```shell script
$ curl -X POST \
      -H "content-type: application/json"  \
      -H "ce-specversion: 1.0"  \
      -H "ce-source: /hospital/entry"  \
      -H "ce-type: new.patients.events"  \
      -H "ce-id: 12346"  \
      -d "{ \"name\": \"Mick\", \"dateOfBirth\": \"1983-08-15\", \"symptoms\":[\"seizures\"]}" \
  http://localhost:8080
```

> In the application log, you will see something like:

Reviewer 2: ```suggestion
In the application's log, you will see something like:
```

```

Then query the appointments' resource to see the assigned doctor:

```shell script
$ curl http://localhost:8080/onboarding/schedule/appointment | jq '.'

[
  {
    "doctor": {
      "id": "8293dc94-2386-11eb-adc1-0242ac120002",
      "name": "Maria Mind",
      "specialty": "Neurology"
    },
    "patient": {
      "name": "Mick",
      "id": "5560ed3d-becb-4f35-8c6a-2816338bf8fd",
      "symptoms": [
        "seizures"
      ],
      "dateOfBirth": "1983-08-15",
      "assignedDoctor": {
        "id": "8293dc94-2386-11eb-adc1-0242ac120002",
        "name": "Maria Mind",
        "specialty": "Neurology"
      }
    },
    "date": "2020-11-18 09:00"
  }
]
```

_**Note:** Please make sure you have [jq](https://stedolan.github.io/jq/download/) installed in your system to see the formatted output._

### Compile and Run in JVM mode

```text
mvn clean package 
java -jar target/serverless-workflow-functions-quarkus-runner.jar   
```

or on windows

```text
mvn clean package
java -jar target\serverless-workflow-functions-quarkus-runner.jar
```

### Compile and Run using Local Native Image

> Note that this requires GRAALVM_HOME to point to a valid GraalVM installation

Reviewer 2: See comment above

```text
mvn clean package -Pnative
```
  
To run the generated native executable, generated in `target/`, execute

```text
./target/serverless-workflow-functions-quarkus-runner.jar
```

## Deploying with Kogito Operator

> In the [`operator`](operator) directory you'll find the custom resources needed to deploy this example on OpenShift or Kubernetes with the [Kogito Operator](https://docs.jboss.org/kogito/release/latest/html_single/#chap_kogito-deploying-on-openshift).

Reviewer 2: ```suggestion
In the [`operator`](operator) directory you can find the custom resources needed to deploy this example on OpenShift or Kubernetes with the [Kogito Operator](https://docs.jboss.org/kogito/release/latest/html_single/#chap_kogito-deploying-on-openshift).
```

> 

Reviewer 2: ```suggestion

Don't forget to replace the `<namespace>` placeholder in yaml files

```
